### ISSUE TITLE
Sudden tight loop on connection problems

### ISSUE THREAD

**Issue by dvv** [codetheweb/tuyapi#399]
**Describe the bug**

The system goes to 100% CPU usage, TUYA devices become non-controllable.

**To Reproduce**

Wait enough running tuya-mqtt gate.

**Expected behavior**

- Either exit the process so that supervisor could restart it, or:

- Properly guard https://github.com/codetheweb/tuyapi/blob/master/index.js#L138 with chainable `.catch()` so that exception thrown here https://github.com/codetheweb/tuyapi/blob/master/index.js#L275 be caught and passed to the caller.

**Debug Output**

```
Jan 03 18:43:27 gate 09270ff75f9d[473]: (node:1) UnhandledPromiseRejectionWarning: Error: No connection has been made to the device.                                          
Jan 03 18:43:27 gate 09270ff75f9d[473]:     at TuyaDevice._send (/app/node_modules/tuyapi/index.js:275:13)                                                                    
Jan 03 18:43:27 gate 09270ff75f9d[473]:     at /app/node_modules/tuyapi/index.js:138:14                                                                                       
Jan 03 18:43:27 gate 09270ff75f9d[473]:     at new Promise (<anonymous>)                                                                                                      
Jan 03 18:43:27 gate 09270ff75f9d[473]:     at TuyaDevice.get (/app/node_modules/tuyapi/index.js:135:12)                                                                      
Jan 03 18:43:27 gate 09270ff75f9d[473]:     at GenericDevice.init (/app/devices/generic-device.js:15:46)                                                                      
Jan 03 18:43:27 gate 09270ff75f9d[473]:     at republishDevices (/app/tuya-mqtt.js:68:20)                                                                                     
Jan 03 18:43:27 gate 09270ff75f9d[473]: (node:1) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `-
-unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 3)                                                          
Jan 03 18:43:27 gate 09270ff75f9d[473]: (node:1) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.                                                                                                            
```

**Screenshots**

**Desktop (please complete the following information):**
 - OS: Linux gate 5.8.9-1-ARCH #1 SMP PREEMPT Sat Sep 12 20:20:55 UTC 2020 armv7l GNU/Linux
 - OS Version 5.8.9-1-ARCH
 - Node Version v14.14.0

**Additional context**

**Comment by codetheweb** [comment]
Thanks for the report; that code is indeed very wrong.

Published a fix in v6.1.2, feel free to reopen if it's not fixed.

**Comment by dvv** [comment]
Thank you for the quick fix!

I had a power shortage a bunch of minutes ago (tuya devices became down then up) and after that issuing a toggle led to:

```
/home/vd/Downloads/tuya/node_modules/tuyapi/index.js:272
      throw new Error('No connection has been made to the device.');
            ^

Error: No connection has been made to the device.
    at TuyaDevice._send (/home/vd/Downloads/tuya/node_modules/tuyapi/index.js:272:13)
    at /home/vd/Downloads/tuya/node_modules/tuyapi/index.js:137:12
    at new Promise (<anonymous>)
    at TuyaDevice.get (/home/vd/Downloads/tuya/node_modules/tuyapi/index.js:135:12)
    at TuyaDevice.toggle (/home/vd/Downloads/tuya/node_modules/tuyapi/index.js:679:31)
```

Indeed, despite description at https://github.com/codetheweb/tuyapi/blob/master/index.js#L262-L264 , `_send` helper does not queue a reconnect (AFAICS) and barely calling `get` inside the `toggle` chokes.

Is this the expected behavior or shouldn't `get` be internally wrapped as `set` is https://github.com/codetheweb/tuyapi/blob/master/index.js#L241 ?

TIA

**Comment by codetheweb** [comment]
> Indeed, despite description at https://github.com/codetheweb/tuyapi/blob/master/index.js#L262-L264 , _send helper does not queue a reconnect (AFAICS) and barely calling get inside the toggle chokes.

Yeah, I think that comment is out of date. Adding `await this.connect()` right before https://github.com/codetheweb/tuyapi/blob/46bb3739202150097185457e0690c7e281db6db9/index.js#L279 would probably make sense?

> Is this the expected behavior or shouldn't get be internally wrapped as set is https://github.com/codetheweb/tuyapi/blob/master/index.js#L241 ?

It's been a while since I've touched that code. The queue is in place for `set()` so that race conditions don't occur. I suppose it's possible for race conditions to occur when calling `get()` as well.

**Comment by dvv** [comment]
I believe `await this.connect()` there would be right.
And https://github.com/codetheweb/tuyapi/blob/master/index.js#L270-L273 should be fired in form of `reject` only if the last retry failed, imho.
Can you push the changes? TIA

**Comment by codetheweb** [comment]
> And https://github.com/codetheweb/tuyapi/blob/master/index.js#L270-L273 should be fired in form of reject only if the last retry failed, imho.

This is the default behavior of the `p-retry` package.

Published a fix in v6.1.3.

**Comment by dvv** [comment]
Thank you! I believe we're done with this.
